src/factories/flowRunArtifacts.ts

```diff
@@ -0,0 +1,233 @@
+import throttle from 'lodash.throttle'
+import { Container } from 'pixi.js'
+import { DEFAULT_ROOT_ARTIFACT_COLLISION_THROTTLE, DEFAULT_ROOT_ARTIFACT_Z_INDEX } from '@/consts'
+import { ArtifactFactory } from '@/factories/artifact'
+import { ArtifactClusterFactory } from '@/factories/artifactCluster'
+import { flowRunArtifactFactory, FlowRunArtifactFactory } from '@/factories/flowRunArtifact'
+import { Artifact } from '@/models'
+import { BoundsContainer } from '@/models/boundsContainer'
+import { waitForApplication, waitForViewport } from '@/objects'
+import { waitForConfig } from '@/objects/config'
+import { emitter } from '@/objects/events'
+import { layout, waitForSettings } from '@/objects/settings'
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export async function flowRunArtifactsFactory() {
+  const application = await waitForApplication()
+  const viewport = await waitForViewport()
+  const config = await waitForConfig()
+  const settings = await waitForSettings()
+
+  const artifacts: Map<string, ArtifactFactory> = new Map()
+  const clusterNodes: ArtifactClusterFactory[] = []
+
+  let container: Container | null = null
+  let internalData: Artifact[] | null = null
+  let nonTemporalAlignmentEngaged = false
+
+  emitter.on('viewportMoved', () => update())
+
+  async function render(newData?: Artifact[]): Promise<void> {
+    if (container) {
+      container.visible = !settings.disableArtifacts
+    }
+
+    if (settings.disableArtifacts) {
+      return
+    }
+
+    if (newData) {
+      internalData = newData
+    }
+
+    if (!internalData) {
+      return
+    }
+
+    if (!container) {
+      createContainer()
+    }
+
+    const promises: Promise<BoundsContainer>[] = []
+
+    for (const artifact of internalData) {
+      promises.push(createArtifact(artifact))
+    }
+
+    await Promise.all(promises)
+  }
+
+  async function createArtifact(artifact: Artifact): Promise<BoundsContainer> {
+    if (artifacts.has(artifact.id)) {
+      return artifacts.get(artifact.id)!.render()
+    }
+
+    const factory = await flowRunArtifactFactory({ type: 'artifact', artifact }) as ArtifactFactory
```

Comment:
Why do we need an `as` here?

Reply:
I updated the type definitions in [flowRunArtifact](https://github.com/PrefectHQ/graphs/pull/409/files#diff-86da312407c4c0eb549c66a3bf11bf81d2f1b720d17b6231b80a4f9c8ee432aaR15) so these `as` assignments are no longer here (very happy to remove that need). Could I ask you to sign off on the implementation there?

Reply:
Nice, yeah that's a nice change. Only thought is does it work to add that new factory type to the `getFactory` function so that you don't even need the `as` on the return? If not no worries, improvement either way.

Reply:
It appears typescript can't narrow the type of T within the if statement there. I messed around with different ways of doing it, including a type guard for the condition. Learned some stuff about typescript getting this far though.